tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1466,22 +1475,44 @@ SpirvVariable *DeclResultIdMapper::createPushConstant(const VarDecl *decl) {
 SpirvVariable *
 DeclResultIdMapper::createShaderRecordBuffer(const VarDecl *decl,
                                              ContextUsageKind kind) {
+  const QualType type = decl->getType();
   const auto *recordType =
-      hlsl::GetHLSLResourceResultType(decl->getType())->getAs<RecordType>();
+      hlsl::GetHLSLResourceResultType(type)->getAs<RecordType>();
   assert(recordType);
 
   assert(kind == ContextUsageKind::ShaderRecordBufferEXT ||
          kind == ContextUsageKind::ShaderRecordBufferNV);
 
-  const auto typeName = kind == ContextUsageKind::ShaderRecordBufferEXT
-                            ? "type.ShaderRecordBufferEXT."
-                            : "type.ShaderRecordBufferNV.";
+  SpirvVariable *var;
+  if (isConstantBuffer(type)) {
+    // Constant buffers already have Block decoration. The variable will need
+    // the appropriate storage class.
+
+    const auto sc = kind == ContextUsageKind::ShaderRecordBufferNV
+                        ? spv::StorageClass::ShaderRecordBufferNV
+                        : spv::StorageClass::ShaderRecordBufferKHR;
+
+    // Create the variable for the whole struct / struct array.
+    // The fields may be 'precise', but the structure itself is not.
+    var = spvBuilder.addModuleVar(type, sc,
+                                  /*isPrecise*/ false,
+                                  /*isNoInterp*/ false, decl->getName());
+
+    const SpirvLayoutRule layoutRule = spirvOptions.sBufferLayoutRule;
 
-  const std::string structName =
-      typeName + recordType->getDecl()->getName().str();
-  SpirvVariable *var = createStructOrStructArrayVarOfExplicitLayout(
-      recordType->getDecl(), /*arraySize*/ 0, kind, structName,
-      decl->getName());
+    var->setHlslUserType("");
+    var->setLayoutRule(layoutRule);
+  } else {
+    const auto typeName = kind == ContextUsageKind::ShaderRecordBufferEXT
+                              ? "type.ShaderRecordBufferEXT."
```

Comment:
nit: name is EXT, but storage class is now KHR

Reply:
I've changed the typename and the enum identifier.